Sources/Action.swift

```diff
@@ -80,39 +91,103 @@ public final class Action<Input, Output, Error: Swift.Error> {
 	///
 	/// - parameters:
 	///   - state: A property to be the state of the `Action`.
-	///   - enabledIf: A predicate which determines the availability of the `Action`,
+	///   - isEnabled: A predicate which determines the availability of the `Action`,
 	///                given the latest `Action` state.
 	///   - execute: A closure that produces a unit of work, as `SignalProducer`, to be
 	///              executed by the `Action`.
-	public init<State: PropertyProtocol>(state property: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+	public init<State: PropertyProtocol>(state: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+		typealias ActionState = (availability: Availability, value: State.Value)
+		let isUserEnabled = isEnabled
+
 		deinitToken = Lifetime.Token()
 		lifetime = Lifetime(deinitToken)
-		
-		// Retain the `property` for the created `Action`.
-		lifetime.observeEnded { _ = property }
 
-		executeClosure = { state, input in execute(state as! State.Value, input) }
+		// `isExecuting` and `isEnabled` should use their own locks, so that legitimate
+		// feedbacks would not deadlock.
+		let initial = (availability: Availability.enabledIdle, value: state.value)
+		let actionState = MutableProperty<ActionState>(initial)
+		let isExecuting = MutableProperty(false)
+		let isEnabled = MutableProperty(true)
+		self.isExecuting = Property(capturing: isExecuting)
+		self.isEnabled = Property(capturing: isEnabled)
```

Comment:
The signal of `ActionState` is deliberately ignored, since `Signal` doesn't support recursion with value events. 

`Signal` deadlocks when one do `isEnabled <~ isExecuting` and both are derived from the same `MutableProperty`, i.e. the current implementation.